jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -128,21 +129,42 @@ object XGBoost extends Serializable {
       }
       rabitEnv.put("DMLC_TASK_ID", TaskContext.getPartitionId().toString)
       Rabit.init(rabitEnv)
-      val trainingMatrix = new DMatrix(
-        fromDenseToSparseLabeledPoints(trainingPoints, missing), cacheFileName)
+      val trainTestRatio = params.get("trainTestRatio").map(_.toString.toDouble).getOrElse(1.0)
+      // In the worst-case this would store [[trainTestRatio]] of points
+      // buffered in memory.
+      val r = params.get("seed")
+          .map(v => new Random(v.toString.toLong))
+          .getOrElse(new Random())
+      val (trainPoints, testPoints) = fromDenseToSparseLabeledPoints(labeledPoints, missing)
+          .partition(_ => r.nextDouble() <= trainTestRatio)
+      val trainMatrix = new DMatrix(trainPoints, cacheFileName)
+      val testMatrix = new DMatrix(testPoints, cacheFileName)
+      val watches = mutable.LinkedHashMap("train" -> trainMatrix)
+      if (testMatrix.rowNum > 0) {
+        watches += "test" -> testMatrix
+      }
```

Comment:
can we wrap in a helper function? let's try to keep functions around 30 lines....

Reply:
Will do.